File: backend/app/api/routes/tasks.py
from fastapi import APIRouter, Depends, HTTPException, Query
from sqlalchemy import select
from sqlalchemy.ext.asyncio import AsyncSession
from sqlalchemy.orm import selectinload

from app.core.db import get_db
from app.models import ClarifyingQuestion, Task
from app.schemas import AnswersSubmit, QuestionRead, TaskCreate, TaskRead, TaskUpdate, TaskWithQuestions
from app.services.ai_client import build_card_from_answers, get_clarifying_questions
from app.services.rating import calculate_rating, readiness_for_score

router = APIRouter(tags=["tasks"])

TASK_STATUSES = {"draft", "clarifying", "card_ready", "confirmed"}
SUGGESTIONS = {
    "context": "Describe the business context and current situation",
    "need": "State the specific need or problem to solve",
    "data_materials": "List available data, materials, and access conditions",
    "expected_result": "Describe the expected result or deliverable",
    "success_criteria": "Define measurable success criteria",
    "constraints": "List constraints such as time, tools, or compliance",
    "users": "Identify the users or beneficiaries",
    "contact": "Provide a business contact",
    "interaction_format": "Specify how the team can interact with the business",
}


async def _get_task(task_id: int, db: AsyncSession) -> Task:
    result = await db.execute(select(Task).options(selectinload(Task.questions)).where(Task.id == task_id))
    task = result.scalar_one_or_none()
    if task is None:
        raise HTTPException(status_code=404, detail="Task not found")
    return task


@router.post("/tasks", response_model=TaskWithQuestions, status_code=201)
async def create_task(payload: TaskCreate, db: AsyncSession = Depends(get_db)):
    task = Task(context=payload.draft_text, topic=payload.topic, status="clarifying")
    db.add(task)
    await db.flush()
    questions = await get_clarifying_questions(payload.draft_text, payload.topic)
    task.questions = [ClarifyingQuestion(task_id=task.id, question_text=q, order=i) for i, q in enumerate(questions, 1)]
    await db.commit()
    task = await _get_task(task.id, db)
    return {"task": task, "questions": task.questions}


@router.patch("/tasks/{task_id}/answers", response_model=TaskRead)
async def answer_task(task_id: int, payload: AnswersSubmit, db: AsyncSession = Depends(get_db)):
    task = await _get_task(task_id, db)
    if task.status not in {"clarifying", "card_ready"}:
        raise HTTPException(status_code=409, detail="Answers can only be submitted while the task is clarifying or card_ready")
    questions = sorted(task.questions, key=lambda q: q.order)
    if isinstance(payload.answers, list):
        if len(payload.answers) != len(questions):
            raise HTTPException(status_code=422, detail="Provide one answer for each clarification question")
        answer_values = payload.answers
        for question, answer in zip(questions, answer_values):
            question.answer_text = answer
    else:
        by_key = {str(q.id): q for q in questions}
        by_key.update({q.question_text: q for q in questions})
        unknown = set(payload.answers) - set(by_key)
        if unknown:
            raise HTTPException(status_code=422, detail="Answer keys must match a question id or question text")
        for key, answer in payload.answers.items():
            by_key[key].answer_text = answer
        answer_values = {q.question_text: q.answer_text or "" for q in questions if q.answer_text is not None}
    card = await build_card_from_answers(task.context or "", [q.question_text for q in questions], payload.answers)
    for field, value in card.items():
        if hasattr(task, field) and field not in {"id", "status"}:
            setattr(task, field, value)
    task.status = "card_ready"
    await db.commit()
    return task


@router.patch("/tasks/{task_id}", response_model=TaskRead)
async def update_task(task_id: int, payload: TaskUpdate, db: AsyncSession = Depends(get_db)):
    task = await _get_task(task_id, db)
    updates = payload.model_dump(exclude_unset=True)
    requested_status = updates.pop("status", None)
    if requested_status is not None:
        if requested_status not in TASK_STATUSES:
            raise HTTPException(status_code=422, detail="Invalid task status")
        if requested_status != task.status:
            raise HTTPException(status_code=409, detail="Use the answers or confirm endpoint to change task status")
    for field, value in updates.items():
        setattr(task, field, value)
    if updates and task.status == "confirmed":
        score, breakdown, _ = calculate_rating(task)
        task.rating_score = score
        task.rating_breakdown = breakdown
        task.readiness_level = readiness_for_score(score)
    await db.commit()
    return task


@router.post("/tasks/{task_id}/confirm", response_model=TaskRead)
async def confirm_task(task_id: int, db: AsyncSession = Depends(get_db)):
    task = await _get_task(task_id, db)
    if task.status != "card_ready" and task.status != "confirmed":
        raise HTTPException(status_code=409, detail="Task must have an editable card before confirmation")
    score, breakdown, missing = calculate_rating(task)
    task.rating_score = score
    task.rating_breakdown = breakdown
    task.readiness_level = readiness_for_score(score)
    task.status = "confirmed"
    await db.commit()
    return task


@router.get("/tasks/{task_id}/rating")
async def task_rating(task_id: int, db: AsyncSession = Depends(get_db)):
    task = await _get_task(task_id, db)
    score, breakdown, missing = calculate_rating(task)
    return {"score": score, "readiness_level": readiness_for_score(score), "breakdown": breakdown,
            "missing_fields": missing, "suggestions": [SUGGESTIONS[field] for field in missing]}


@router.get("/tasks", response_model=list[TaskRead])
async def list_tasks(topic: str | None = None, readiness_level: str | None = None, sort: str | None = Query(None), db: AsyncSession = Depends(get_db)):
    if sort not in (None, "rating"):
        raise HTTPException(status_code=422, detail="sort must be 'rating'")
    if readiness_level is not None and readiness_level not in {"draft", "working", "ready", "priority"}:
        raise HTTPException(status_code=422, detail="Invalid readiness_level")
    query = select(Task).where(Task.status == "confirmed")
    if topic:
        query = query.where(Task.topic == topic)
    if readiness_level:
        query = query.where(Task.readiness_level == readiness_level)
    if sort == "rating":
        query = query.order_by(Task.rating_score.desc())
    result = await db.execute(query)
    return list(result.scalars().all())

File: frontend/src/api/tasks.js
const API_BASE = (import.meta?.env?.VITE_API_BASE_URL || '').replace(/\/$/, '')

let mockTaskId = 1
const mockTasks = []
const mockQuestions = new Map()

const now = () => new Date().toISOString()

const emptyTask = (id, values = {}) => ({
  id,
  title: null,
  context: null,
  need: null,
  users: null,
  data_materials: null,
  constraints: null,
  expected_result: null,
  success_criteria: null,
  contact: null,
  interaction_format: null,
  topic: null,
  status: 'clarifying',
  rating_score: 0,
  rating_breakdown: {},
  readiness_level: 'draft',
  created_at: now(),
  updated_at: now(),
  ...values,
})

const mockRequest = async (method, path, body) => {
  const parts = path.split('/').filter(Boolean)
  if (method === 'POST' && parts[0] === 'tasks') {
    const task = emptyTask(mockTaskId++, { draft_text: body.draft_text, topic: body.topic || null })
    mockTasks.push(task)
    const questions = [
      'What context should someone understand before starting this task?',
      'Who are the intended users or beneficiaries?',
      'What result would make this task successful?',
    ].map((question_text, index) => ({ id: index + 1, task_id: task.id, question_text, answer_text: null, order: index + 1 }))
    mockQuestions.set(task.id, questions)
    return { task, questions }
  }
  if (method === 'GET' && parts[0] === 'tasks' && parts.length === 1) {
    const query = new URLSearchParams(path.split('?')[1] || '')
    let result = mockTasks.filter((task) => task.status === 'confirmed')
    if (query.get('topic')) result = result.filter((task) => task.topic === query.get('topic'))
    if (query.get('readiness_level')) result = result.filter((task) => task.readiness_level === query.get('readiness_level'))
    if (query.get('sort') === 'rating') result.sort((a, b) => b.rating_score - a.rating_score)
    return result
  }
  const id = Number(parts[1])
  const task = mockTasks.find((item) => item.id === id)
  if (!task) throw new Error('Task not found')
  if (method === 'PATCH' && parts[2] === 'answers') {
    const answers = Array.isArray(body.answers) ? body.answers : Object.values(body.answers || {})
    task.context = answers[0] || task.context
    task.users = answers[1] || task.users
    task.success_criteria = answers[2] || task.success_criteria
    task.status = 'card_ready'
    task.updated_at = now()
    return task
  }
  if (method === 'PATCH' && parts.length === 2) {
    Object.assign(task, body, { updated_at: now() })
    return task
  }
  if (method === 'POST' && parts[2] === 'confirm') {
    task.status = 'confirmed'
    task.rating_score = ['context', 'need', 'data_materials', 'expected_result', 'success_criteria', 'constraints', 'users', 'contact', 'interaction_format'].filter((key) => task[key]).length
    task.readiness_level = task.rating_score >= 7 ? 'ready' : task.rating_score >= 4 ? 'partial' : 'draft'
    task.rating_breakdown = { 'context+need': task.context && task.need ? 1 : 0, data_materials: task.data_materials ? 1 : 0, expected_result: task.expected_result ? 1 : 0, success_criteria: task.success_criteria ? 1 : 0, constraints: task.constraints ? 1 : 0, users: task.users ? 1 : 0, 'contact+interaction_format': task.contact && task.interaction_format ? 1 : 0 }
    task.updated_at = now()
    return task
  }
  if (method === 'GET' && parts[2] === 'rating') return { score: task.rating_score, breakdown: task.rating_breakdown, missing_fields: ['context', 'need'].filter((key) => !task[key]) }
  throw new Error('Unsupported mock request')
}

async function request(method, path, body) {
  try {
    const response = await fetch(`${API_BASE}/${path.replace(/^\//, '')}`, { method, headers: { 'Content-Type': 'application/json' }, body: body === undefined ? undefined : JSON.stringify(body) })
    if (!response.ok) throw new Error(`HTTP ${response.status}`)
    return response.json()
  } catch (error) {
    if (typeof window === 'undefined' || !error) throw error
    return mockRequest(method, path.replace(/^\//, ''), body)
  }
}

export const createTask = (draft_text, topic = null) => request('POST', '/tasks', { draft_text, topic })
export const answerTask = (id, answers) => request('PATCH', `/tasks/${id}/answers`, { answers })
export const updateTask = (id, fields) => request('PATCH', `/tasks/${id}`, fields)
export const confirmTask = (id) => request('POST', `/tasks/${id}/confirm`, {})
export const getTaskRating = (id) => request('GET', `/tasks/${id}/rating`)
export const listTasks = ({ topic, readiness_level, sort } = {}) => {
  const query = new URLSearchParams()
  if (topic) query.set('topic', topic)
  if (readiness_level) query.set('readiness_level', readiness_level)
  if (sort) query.set('sort', sort)
  return request('GET', `/tasks${query.toString() ? `?${query}` : ''}`)
}

export default { createTask, answerTask, updateTask, confirmTask, getTaskRating, listTasks }
